File: 56-RADStudio10.3/AndroidAPILevelUpgrade_Demos/CPP/Multi-Device Samples/Device Sensors and Services/Bluetooth/BLEScanner/Unit6.cpp
//---------------------------------------------------------------------------
// Copyright (c) 2016 Embarcadero Technologies, Inc. All rights reserved.  
//
// This software is the copyrighted property of Embarcadero Technologies, Inc. 
// ("Embarcadero") and its licensors. You may only use this software if you 
// are an authorized licensee of Delphi, C++Builder or RAD Studio 
// (the "Embarcadero Products").  This software is subject to Embarcadero's 
// standard software license and support agreement that accompanied your 
// purchase of the Embarcadero Products and is considered a Redistributable, 
// as such term is defined thereunder. Your use of this software constitutes 
// your acknowledgement of your agreement to the foregoing software license 
// and support agreement. 
//---------------------------------------------------------------------------
#include <fmx.h>
#ifdef __ANDROID__
    #include <Androidapi.Helpers.hpp>
    #include <Androidapi.JNI.Os.hpp>
#endif
#include <FMX.DialogService.hpp>
#pragma hdrstop

#include "Unit6.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.fmx"
TForm6 *Form6;

class TTPDiscoverServices : public TCppInterfacedObject<TProc>
{
  public:
	TTPDiscoverServices(TForm6 *AForm) { FForm = AForm; }
	void __fastcall Invoke();
  private:
	TForm6 *FForm;
	void __fastcall SynchronizeProc();
};

void __fastcall TTPDiscoverServices::SynchronizeProc(void)
{
	FForm->ListBox2->Items->Add("- Service discovery not allowed");
	FForm->ListBox1->Enabled = true;
}

void __fastcall TTPDiscoverServices::Invoke(void)
{
	if(!FForm->BluetoothLE1->DiscoveredDevices->Items[FForm->ListBox1->ItemIndex]->DiscoverServices())
		TThread::Synchronize(NULL, SynchronizeProc);
}

//---------------------------------------------------------------------------
__fastcall TForm6::TForm6(TComponent* Owner)
	: TForm(Owner)
{
#ifdef __ANDROID__
    FLocationPermission = JStringToString(TJManifest_permission::JavaClass->ACCESS_COARSE_LOCATION);
#endif
}
//---------------------------------------------------------------------------
void __fastcall TForm6::DisplayRationale(TObject* Sender, const TStringDynArray APermissions, const _di_TProc APostRationaleProc)
{
    // Show an explanation to the user *asynchronously* - don't block this thread waiting for the user's response!
    // After the user sees the explanation, invoke the post-rationale routine to request the permissions
    TDialogService::ShowMessage("We need to be given permission to discover BLE devices",
        [APostRationaleProc](TModalResult AKey)
        {
            APostRationaleProc->Invoke();
        });
}
//---------------------------------------------------------------------------
void __fastcall TForm6::RequestPermissionsResult(TObject* Sender, const System::TArray__1<String> APermissions, const System::TArray__1<TPermissionStatus> AGrantResults)
{
    // 1 permission involved: ACCESS_COARSE_LOCATION
	if ((AGrantResults.Length == 1) && (AGrantResults[0] == TPermissionStatus::Granted))
		StartBLEDiscovery();
	else
		ShowMessage("Cannot start BLE scan as the permission was not granted");
}
//---------------------------------------------------------------------------
void __fastcall TForm6::btnStartScanClick(TObject *Sender)
{
    TStringDynArray perms;
    perms.set_length(1);
    perms[0] = FLocationPermission;
    PermissionsService()->RequestPermissions(perms, RequestPermissionsResult, DisplayRationale);
}
//---------------------------------------------------------------------------
void __fastcall TForm6::btnStopScanClick(TObject *Sender)
{
    StopBLEDiscovery();
}
//---------------------------------------------------------------------------
void TForm6::StartBLEDiscovery()
{
	if(!Scanning) {
		ListBox1->Clear();
		ScanningStart = TThread::GetTickCount();
		BluetoothLE1->DiscoverDevices(ScanningTime);
		ProgressBar1->Value = 0;
		Timer1->Enabled = true;
		Scanning = true;
	}
}
//---------------------------------------------------------------------------
void TForm6::StopBLEDiscovery()
{
	Timer1->Enabled = false;
	Scanning = false;
	BluetoothLE1->CancelDiscovery();
}
//---------------------------------------------------------------------------
void __fastcall TForm6::FormShow(TObject *Sender)
{
	Scanning = false;
}
//---------------------------------------------------------------------------
void __fastcall TForm6::Timer1Timer(TObject *Sender)
{
	DWORD LElapsed = TThread::GetTickCount() - ScanningStart;
	ProgressBar1->Value = ProgressBar1->Max * float(LElapsed)/ScanningTime;
}
//---------------------------------------------------------------------------
void __fastcall TForm6::ListBox1ItemClick(TCustomListBox * const Sender, TListBoxItem * const Item)
{
	btnStopScanClick(btnStopScan);
	ListBox2->Clear();
	ListBox2->Items->Add("- Discovering services -->");

	TThread::CreateAnonymousThread(new TTPDiscoverServices(this))->Start();

	ListBox1->Enabled = false;
}
//---------------------------------------------------------------------------
void __fastcall TForm6::BluetoothLE1DiscoverLEDevice(TObject * const Sender, TBluetoothLEDevice * const ADevice,
		  int Rssi, TScanResponse * const ScanResponse)
{
	int dCount = BluetoothLE1->DiscoveredDevices->Count;
	int numOfDevices = ListBox1->Count;
	String Name;
	for (int i = 0; i < dCount; i++) {
		Name = BluetoothLE1->DiscoveredDevices->Items[i]->DeviceName;
		if(Name == "")
			Name = "Unknown device";
		Name = IntToStr(i+1) + L" - " + Name + L" - " + BluetoothLE1->DiscoveredDevices->Items[i]->Identifier();
		if(numOfDevices == i) {
			 ListBox1->Items->Add(Name);
		}
		else {
			ListBox1->Items->Strings[i] = Name;
		}
	}
}
//---------------------------------------------------------------------------

void __fastcall TForm6::BluetoothLE1EndDiscoverDevices(TObject * const Sender, TBluetoothLEDeviceList * const ADeviceList)
{
	if (Scanning) {
		ProgressBar1->Value = ProgressBar1->Max;
	}
	Timer1->Enabled = false;
	Scanning = false;
}
//---------------------------------------------------------------------------

void __fastcall TForm6::BluetoothLE1ServicesDiscovered(TObject * const Sender, TBluetoothGattServiceList * const AServiceList)
{
	String aux;
	if (AServiceList->Count > 0) {
		for (int i = 0; i < AServiceList->Count; i++) {
			TBluetoothGattService *service = AServiceList->Items[i];
			aux = IntToStr(i+1) + L" - " + service->UUIDName + L" - " + GUIDToString(service->UUID);
			ListBox2->Items->Add(aux);
			for (int c = 0; c < service->Characteristics->Count; c++) {
				TBluetoothGattCharacteristic *chara = service->Characteristics->Items[c];
				aux = L"    - " + chara->UUIDName + L" - " + GUIDToString(chara->UUID);
				ListBox2->Items->Add(aux);
			}
        }
	}
	else {
		ListBox2->Items->Add("- Access not allowed or no services");
	}
	ListBox1->Enabled = true;
}
//---------------------------------------------------------------------------

File: 56-RADStudio10.3/AndroidAPILevelUpgrade_Demos/Object Pascal/Multi-Device Samples/Device Sensors and Services/Bluetooth/BLEScanner/Unit6.pas
//---------------------------------------------------------------------------

// This software is Copyright (c) 2015 Embarcadero Technologies, Inc.
// You may only use this software if you are an authorized licensee
// of an Embarcadero developer tools product.
// This software is considered a Redistributable as defined under
// the software license agreement that comes with the Embarcadero Products
// and is subject to that software license agreement.

//---------------------------------------------------------------------------

unit Unit6;

interface

uses
  System.SysUtils, System.Types, System.UITypes, System.Classes, System.Variants,
  FMX.Types, FMX.Controls, FMX.Forms, FMX.Graphics, FMX.Dialogs,
  System.Bluetooth, FMX.StdCtrls, System.Bluetooth.Components, System.Permissions,
  FMX.Layouts, FMX.ListBox, FMX.Controls.Presentation;

type
  TForm6 = class(TForm)
    Panel1: TPanel;
    ListBox1: TListBox;
    BluetoothLE1: TBluetoothLE;
    btnStartScan: TButton;
    btnStopScan: TButton;
    Timer1: TTimer;
    ProgressBar1: TProgressBar;
    ListBox2: TListBox;
    procedure btnStartScanClick(Sender: TObject);
    procedure btnStopScanClick(Sender: TObject);
    procedure BluetoothLE1DiscoverLEDevice(const Sender: TObject; const ADevice: TBluetoothLEDevice; Rssi: Integer; const ScanResponse: TScanResponse);
    procedure Timer1Timer(Sender: TObject);
    procedure BluetoothLE1EndDiscoverDevices(const Sender: TObject; const ADeviceList: TBluetoothLEDeviceList);
    procedure FormShow(Sender: TObject);
    procedure BluetoothLE1ServicesDiscovered(const Sender: TObject; const AServiceList: TBluetoothGattServiceList);
    procedure ListBox1ItemClick(const Sender: TCustomListBox; const Item: TListBoxItem);
    procedure FormCreate(Sender: TObject);
  private
    { Private declarations }
    Scanning: Boolean;
    ScanningStart: Cardinal;
    FLocationPermission: string;
    procedure RequestPermissionsResult(Sender: TObject; const APermissions: TArray<string>; const AGrantResults: TArray<TPermissionStatus>);
    procedure DisplayRationale(Sender: TObject; const APermissions: TArray<string>; const APostRationaleProc: TProc);
    procedure StartBLEDiscovery;
    procedure StopBLEDiscovery;
  public
    { Public declarations }
  end;

const
  ScanningTime = 10000; // 10s in msecs

var
  Form6: TForm6;

implementation

uses
{$IFDEF ANDROID}
  Androidapi.Helpers,
  Androidapi.JNI.JavaTypes,
  Androidapi.JNI.Os,
{$ENDIF}
  FMX.DialogService;

{$R *.fmx}

procedure TForm6.BluetoothLE1DiscoverLEDevice(const Sender: TObject; const ADevice: TBluetoothLEDevice; Rssi: Integer; const ScanResponse: TScanResponse);
var
  PrevDiscoveredDevicesCount: Integer;
  DiscoveredDevicesCount: Integer;
  DiscoveredDeviceIndex: Integer;
  DiscoveredDevice: TBluetoothLEDevice;
  DiscoveredDeviceName: string;
begin
  PrevDiscoveredDevicesCount := Listbox1.Count;
  DiscoveredDevicesCount := BluetoothLE1.DiscoveredDevices.Count;

  for DiscoveredDeviceIndex := 0 to DiscoveredDevicesCount - 1 do
  begin
    DiscoveredDevice := BluetoothLE1.DiscoveredDevices[DiscoveredDeviceIndex];
    DiscoveredDeviceName := DiscoveredDevice.DeviceName;
    if DiscoveredDeviceName = '' then
      DiscoveredDeviceName := 'Unknown device';
    DiscoveredDeviceName := (DiscoveredDeviceIndex + 1).ToString + ' - ' + DiscoveredDeviceName + ' - ' + DiscoveredDevice.Identifier;

    if DiscoveredDeviceIndex = PrevDiscoveredDevicesCount then
      Listbox1.Items.Add(DiscoveredDeviceName)
    else
      Listbox1.Items[DiscoveredDeviceIndex] := DiscoveredDeviceName;
  end;
end;

procedure TForm6.BluetoothLE1EndDiscoverDevices(const Sender: TObject; const ADeviceList: TBluetoothLEDeviceList);
begin
  if Scanning then
    ProgressBar1.Value := ProgressBar1.Max;
  Timer1.Enabled := False;
  Scanning := False;
end;

procedure TForm6.BluetoothLE1ServicesDiscovered(const Sender: TObject; const AServiceList: TBluetoothGattServiceList);
var
  ServiceIndex: Integer;
  Service: TBluetoothGattService;
  CharacteristicIndex: Integer;
  Characteristic: TBluetoothGattCharacteristic;
begin
  if AServiceList.Count > 0 then
  begin
    for ServiceIndex := 0 to AServiceList.Count - 1 do
    begin
      Service := AServiceList[ServiceIndex];
      Listbox2.Items.Add((ServiceIndex + 1).ToString + ' - ' + Service.UUIDName + ' - ' + Service.UUID.ToString);

      for CharacteristicIndex := 0 to Service.Characteristics.Count - 1 do
      begin
        Characteristic := Service.Characteristics[CharacteristicIndex];
        Listbox2.Items.Add('    - ' + Characteristic.UUIDName + ' - ' + Characteristic.UUID.ToString);
      end;
    end;
  end
  else
    Listbox2.Items.Add('- Access not allowed or no service available');

  Listbox1.Enabled := True;
end;

procedure TForm6.btnStartScanClick(Sender: TObject);
begin
  PermissionsService.RequestPermissions([FLocationPermission], RequestPermissionsResult, DisplayRationale);
end;

procedure TForm6.btnStopScanClick(Sender: TObject);
begin
  StopBLEDiscovery
end;

procedure TForm6.FormCreate(Sender: TObject);
begin
{$IFDEF ANDROID}
  FLocationPermission := JStringToString(TJManifest_permission.JavaClass.ACCESS_COARSE_LOCATION);
{$ENDIF}
end;

procedure TForm6.FormShow(Sender: TObject);
begin
  Scanning := False;
end;

procedure TForm6.ListBox1ItemClick(const Sender: TCustomListBox; const Item: TListBoxItem);
begin
  btnStopScanClick(Sender);
  Listbox2.Clear;
  Listbox2.Items.Add('- Discovering services -->');
  TThread.CreateAnonymousThread(
    procedure
    begin
      if not BluetoothLE1.DiscoveredDevices[ListBox1.ItemIndex].DiscoverServices then
        TThread.Synchronize(nil,
          procedure
          begin
            Listbox2.Items.Add('- Service discovery not allowed');
            Listbox1.Enabled := True;
          end);
    end).Start;
  Listbox1.Enabled := False;
end;

procedure TForm6.DisplayRationale(Sender: TObject; const APermissions: TArray<string>; const APostRationaleProc: TProc);
begin
  TDialogService.ShowMessage('We need to be given permission to discover BLE devices',
    procedure(const AResult: TModalResult)
    begin
      APostRationaleProc;
    end)
end;

procedure TForm6.RequestPermissionsResult(Sender: TObject; const APermissions: TArray<string>; const AGrantResults: TArray<TPermissionStatus>);
begin
  // 1 permissions involved: ACCESS_COARSE_LOCATION
  if (Length(AGrantResults) = 1) and (AGrantResults[0] = TPermissionStatus.Granted) then
    StartBLEDiscovery
  else
    TDialogService.ShowMessage('Cannot start BLE scan as the permission has not been granted');
end;

procedure TForm6.StartBLEDiscovery;
begin
  if not Scanning then
  begin
    Listbox1.Clear;
    ScanningStart := TThread.GetTickCount;
    BluetoothLE1.DiscoverDevices(ScanningTime);
    ProgressBar1.Value := 0;
    Timer1.Enabled := True;
    Scanning := True;
  end;
end;

procedure TForm6.StopBLEDiscovery;
begin
  Timer1.Enabled := False;
  Scanning := False;
  BluetoothLE1.CancelDiscovery;
end;

procedure TForm6.Timer1Timer(Sender: TObject);
var
  LElapsed: Cardinal;
begin
  LElapsed := TThread.GetTickCount - ScanningStart;
  ProgressBar1.Value := ProgressBar1.Max * (LElapsed.ToSingle / ScanningTime);
end;

end.
